### ISSUE TITLE
improvement ideas

### ISSUE THREAD

**Issue by denismaciel** [lervag/apy#52]
I really, really like this project and would like to contribute to it. It's such a simple tool, but it has revolutionized the way I interact with Anki.

I have some ideas about what to improve, but would like to get your feedback first, @lervag.

* Add type hints and type-check the code with MyPy.
* Auto-format the code with Black
* Run tests automatically with Github Actions

Do you think these make sense? Let me know if you have other ideas.

Looking forward to contributing!

**Comment by ckp95** [comment]
For my sins, I learned a bunch about Python package management esoterica at my last job. Right now, apy's dependencies aren't fully specified. I can revamp the dependency metadata and build process if you want.

**Comment by lervag** [comment]
> For my sins,

I assume you are referring to the post you deleted and moved to a new issue, #54?

> I learned a bunch about Python package management esoterica at my last job. Right now, apy's dependencies aren't fully specified and the installation process isn't totally clear for various platforms. I can revamp the dependency metadata and build process if you want. Basically make the thing use PEP 518 pyproject.toml and use poetry as the build tool. I've found it's more reliable than hand-rolled setup.py and requirements.txt.

Yes, I would be more than happy to accept PRs to this end!

**Comment by ckp95** [comment]
>I assume you are referring to the post you deleted and moved to a new issue, https://github.com/lervag/apy/issues/54?

Ha, [no](https://en.wiktionary.org/wiki/for_my_sins), just that to get anything done at work I had to turn my brain into a landfill of Python packaging/build system/importlib trivia. And it made me wonder if I'd done something wrong in a past life.

>Yes, I would be more than happy to accept PRs to this end!

Great, I have one in the works right now.

**Comment by lervag** [comment]
> > I assume you are referring to the post you deleted and moved to a new issue, #54?
> 
> Ha, [no](https://en.wiktionary.org/wiki/for_my_sins), just that to get anything done at work I had to turn my brain into a landfill of Python packaging/build system/importlib trivia. And it made me wonder if I'd done something wrong in a past life.

Thanks; I have not seen this phrase used like this before. TIL :)

> > Yes, I would be more than happy to accept PRs to this end!
> 
> Great, I have one in the works right now.
> 
> EDIT: here #55

<3

**Comment by denismaciel** [comment]
Chiming in again here:

Installing `apy` is hard indeed.

Here's the commands I ran the last time I installed it on Ubuntu 22.04. 

```sh
python3.11 -m venv ~/venvs/apy
~/venvs/apy/bin/pip install aqt==2.1.53
~/venvs/apy/bin/pip install git+https://github.com/lervag/apy.git#egg=apy
~/venvs/apy/bin/pip install pyqt5
~/venvs/apy/bin/pip install PyQtWebEngine
~/venvs/apy/bin/pip install html5lib # for apy list
~/venvs/apy/bin/apy
```

The idea of using pipx is really neat. I hope @ckp95 can pull that off.

**Comment by lervag** [comment]
First, thanks to @denismaciel for bringing GitHub Actions and formatting with `black` in #56 - I think it can be merged quite soon after some minor discussions!

From the previous discussion, it seems the only suggestion that is not addressed yet is to add **type hints**. :)

@ckp95

> Chiming in again here:
> 
> Installing `apy` is hard indeed.

For me it has never really been hard. Perhaps by chance, I don't know. I've usually installed it with `pip install -e .` or `pip install .`, and it has "just worked". But I'm aware this is not optimal!

I would not mind learning better ways! I really want two things:

* To have `apy` globally available for my daily/weekly use.
* To have a decent dev setup where I can make modifications and quickly test on real data when I want to.

The `pip install -e .` has worked well for both of these purposes in the past. But I understand it may have some drawbacks. And using `poetry` and a virtualenv does not really provide a good _global_ version, right?

> Here's the commands I ran the last time I installed it on Ubuntu 22.04.

After your contribution in #55, I would have assumed it should suffice to do

```sh
python -m venv ~/venvs/apy
cd /path/to/apy
~/venvs/apy/bin/pip install .
~/venvs/apy/bin/apy
```

Or am I wrong?

> The idea of using pipx is really neat. I hope @ckp95 can pull that off.

I still have no experience wiht `pipx` and would be happy to learn more about it!

**Comment by ckp95** [comment]
>For me it has never really been hard. Perhaps by chance, I don't know. I've usually installed it with pip install -e . or pip install ., and it has "just worked". But I'm aware this is not optimal!
>
>I would not mind learning better ways! I really want two things:
>
>    To have apy globally available for my daily/weekly use.
>    To have a decent dev setup where I can make modifications and quickly test on real data when I want to.
>
>The pip install -e . has worked well for both of these purposes in the past. But I understand it may have some drawbacks. And using poetry and a virtualenv does not really provide a good global version, right?

Right, if you install using the system `python` it may work for a time, but later package upgrades may break something, because your system package manager does not know about `apy` and the transitive dependencies it pulled in. This even applies to `-e` installations, because it's all the same interpreter.

It's this "spooky action at a distance" property that makes Python so infuriating to work with. You can do something wrong but only see symptoms of it weeks or months later, in a completely unrelated program. But doing it the "right" way involves the friction of making and activating a virtual environment. If you read Python development mailing lists, you can see people have been trying to solve this UX problem for literally years, and progress is slow.

`pipx` is one way of getting a globally available install that doesn't interfere with the system Python. But it's not the only way; you can also:

- vendor an entire Python distribution
- make a package that can be installed with the system package manager and put it in a repository like the AUR or something
- ship a docker container / flatpack / snap etc

>I would have assumed it should suffice to do
>
>```
>python -m venv ~/venvs/apy
>cd /path/to/apy
>~/venvs/apy/bin/pip install .
>~/venvs/apy/bin/apy
>```
>
>Or am I wrong?

This should work, let me know if it doesn't.

>I still have no experience wiht pipx and would be happy to learn more about it!

There isn't much to it; if there's a package `foo` for which you can already do `pip install foo`, then `pipx install foo` will automatically create a virtual environment just for that package, and add it to the PATH so that it's available everywhere.

**Comment by lervag** [comment]
> Right, if you install using the system `python` it may work for a time, but later package upgrades may break something, because your system package manager does not know about `apy` and the transitive dependencies it pulled in. This even applies to `-u` installations, because it's all the same interpreter.

Yes, you are right. Although, with experience it becomes easier to handle these situations. I usually resolve issues quite fast. But, again, you are right!

> It's this "spooky action at a distance" property that makes Python projects so infuriating to package and distribute to end-users. You can do something wrong but only see symptoms of it weeks or months later, in a completely unrelated program. But doing it the "right" way involves the friction of making and activating a virtual environment, …

Thanks for the info - an interesting read!

> > I would have assumed it should suffice to do
> > ```
> > python -m venv ~/venvs/apy
> > cd /path/to/apy
> > ~/venvs/apy/bin/pip install .
> > ~/venvs/apy/bin/apy
> > ```
> > Or am I wrong?
> 
> This should work, let me know if it doesn't.

It didn't, actually. We needed to add `pyqt5` or something like that. But after #57 I think it _will_ work. And it seems `poetry install` more or less does exactly the above steps, except running the script (and probably more).

> There isn't much to it; if there's a package `foo` for which you can already do `pip install foo`, then `pipx install foo` will automatically create a virtual environment just for that package, and add it to the PATH so that `foo` is available everywhere.

Cool, that sounds useful. So, `pipx` will ensure that `foo` and `bar` have their separate environments, which would e.g. solve the issue where `foo` and `bar` both depend on _different_ versions of `baz`.

> That does raise the problem of how to get `pipx` in the first place, but I think it's available in most Linux package repositories so you can just `apt install pipx` or `pacman -S pipx` and be okay.

Cool! So, with some time, we should aim for `pipx install apy` and have it "just work"™ :)

(Although, `pip install apy` would also be useful, of course.)

**Comment by ckp95** [comment]
The idea would be: `pipx` for end-users, `poetry` or plain `pip` for development.

**Comment by lervag** [comment]
Yes, that sounds good to me.

**Comment by lervag** [comment]
I believe we've mostly addressed all suggested things in this thread, except adding type hints. If anyone wants to do that work, I will be more than glad to consider a PR! But I don't think I will have the time in any near future.

Can we close this issue?

**Comment by denismaciel** [comment]
> I believe we've mostly addressed all suggested things in this thread, except adding type hints. If anyone wants to do that work, I will be more than glad to consider a PR! But I don't think I will have the time in any near future.
> 
> Can we close this issue?

Yes! I will close the issue and start working on the incremental adoption of mypy soonish.

**Comment by lervag** [comment]
Great! Looking forward to seeing it and to learning more about how to work with mypy :)
